src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt

```diff
@@ -0,0 +1,213 @@
+            .zipWithNext { a, b ->
+                when {
+                    a.labelKey != "title" -> null
+                    b.labelKey != "body" -> constructResource(a, null)
+                    else -> constructResource(a, b)
+                }
+            }
+            .filterNotNull()
+    }
+
+    private fun workbookTake(modelTake: org.wycliffeassociates.otter.common.data.model.Take): Take {
+        // TODO: move into TakeMapper?
+        val workbookTake = Take(
+            name = modelTake.filename,
+            file = modelTake.path,
+            number = modelTake.number,
+            format = MimeType.WAV, // TODO
+            createdTimestamp = modelTake.timestamp,
+            deletedTimestamp = BehaviorRelay.create() // TODO: need DB deleted field
+        )
+        takeMap[workbookTake] = modelTake
+        return workbookTake
+    }
+
+    private fun modelTake(workbookTake: Take): org.wycliffeassociates.otter.common.data.model.Take? {
+        // TODO: move into TakeMapper?
+        return takeMap[workbookTake]
+            ?: org.wycliffeassociates.otter.common.data.model.Take(
+                filename = workbookTake.file.name,
+                path = workbookTake.file,
+                number = workbookTake.number,
+                timestamp = workbookTake.createdTimestamp,
+                played = false,
+                markers = listOf()
+            )
+    }
+
+    private fun constructAssociatedAudio(content: Content): AssociatedAudio {
+        val takes = constructTakes(content)
+        val selectedTake = constructSelectedTake(content)
+        return AssociatedAudio(takes, selectedTake)
+    }
+
+    private fun constructSelectedTake(content: Content): BehaviorRelay<Take?> {
+        // Init the relay with current value from DB
+        val currentSelectedTake = content.selectedTake?.let { workbookTake(it) }
+        val relay = BehaviorRelay.createDefault(currentSelectedTake)
+
+        // Respond to updates pushed into relay by writing them to DB
+        val subscription = relay
+            .distinctUntilChanged() // Don't write unless changed
+            .skip(1) // Don't write the value we just loaded from the DB
+            .subscribe { updatedSelection ->
+                content.selectedTake = updatedSelection?.let { modelTake(it) }
+                contentRepository.update(content)
+            }
+
+        connections += subscription
+        return relay
+    }
+
+    private fun constructTakes(content: Content): ReplayRelay<Take> {
+        val relay = ReplayRelay.create<Take>()
+
+        // Push Takes from DB into relay
+        takeRepository
+            .getByContent(content)
+            .flattenAsObservable { list -> list.sortedBy { it.number } }
+            .map { workbookTake(it) }
+            .subscribe(relay)
+
+        // Respond to new Takes pushed into relay by writing them to DB
+        val subscription = relay.subscribe { newTake ->
```

Comment:
this could get executed as takes are coming out from the database I believe?

Reply:
Yes.  I think that's fixed now.